sysvabi64/sysvabi64.rst

```diff
@@ -916,10 +923,652 @@ These follow the AAELF64_ base definition from Arm.
+function.
+
+The source code interface to an IFUNC resolver is platform
+dependent. The GNU/Linux interface via the GNU C Library is documented
+below.
+
+GNU C Library IFUNC interface
+^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
+
+The GNU C Library (glibc) ``sys/ifunc.h`` header provides the
+necessary type definitions for defining IFUNC resolvers. The resolver
+is passed at least one, and at most 2 parameters dependent on the
+glibc IFUNC ABI version. A flag is set in the first parameter if a
+second parameter is available. The parameters provide access to the
+HWCAP_ flags.
+
+.. code-block:: c
+
+  /* A second argument is passed to the ifunc resolver.  */
+  #define _IFUNC_ARG_HWCAP        (1ULL << 62)
+
+  /* Function prototype for an IFUNC resolver.
+  ElfW(Addr) ifunc_resolver (uint64_t, const __ifunc_arg_t *);
+
+  /* Second argument to an ifunc resolver.  */
+  struct __ifunc_arg_t
+  {
+    unsigned long _size; /* Size of the struct, so it can grow.  */
+    unsigned long _hwcap;
+    unsigned long _hwcap2;
+  };
+
+  typedef struct __ifunc_arg_t __ifunc_arg_t;
+
+IFUNC resolver functions must have a type of ``STT_GNU_IFUNC``. With
+the GCC and Clang compilers an attribute can be used to achieve this.
+
+.. code-block:: c
+
+  /* Implementation alternative 1 */
+  static int implementation1(void) { ... }
+
+  /* Implementation alternative 2 */
+  static int implementation2(void) { ... }
+
+  /* Resolver function */
+  static void* resolver(uint64_t, const __ifunc_arg_t *) { ... }
+
+  /* make symbol ifunc type STT_GNU_IFUNC using resolver as the IFUNC resolver. */
+  int ifunc(void) __attribute__((ifunc("resolver")));
+
+The first ``uint64_t`` parameter is ``AT_HWCAP``. If the
+``_IFUNC_ARG_HWCAP`` flag is set the second parameter is present. The
+second parameter is a structure ``__ifunc_arg_t`` defined in
+``sys/ifunc.h``, it provides access to all of the flags in HWCAP_ via
+fields of the structure.
+
+The IFUNC resolver function returns the address of the function
+implementation.
+
+IFUNC resolvers may be run when the dynamic linker is resolving relocations.
+additional restrictions on what they can contain.
+
+ * IFUNC resolvers must not be compiled with security features like
+   stack-protection, which requires a guard variable to be
+   initialized. Or instrumentation like ASAN that requires a shadow
+   map to be set up.
+
+ * IFUNC resolvers must not have a symbol binding of ``STB_WEAK``.
+
+The order of dynamic relocation resolution across an executable and
+all its shared libraries is platform specific. The following
+recommendations for writing IFUNC resolvers apply to the GNU glibc
+dynamic loader. Other dynamic linkers may have fewer requirements.
+
+ * IFUNC resolver functions must not call a function that may require
+   IFUNC initialization. If the IFUNC initialization for the called
+   function has not occured then undefined behavior results.
+
+ * In position-independent code an IFUNC resolver functions must not
```

Comment:
Tenses `an` IFUNC resolver `functions`   --  I think should either be `an` & `function` or just `functions` without the `an`.

Reply:
Thanks; I've used An IFUNC resolver function.